Cluster upgrade tunning in CAPI (#2307)
giantswarm/docs: src/content/tutorials/fleet-management/cluster-management/cluster-upgrades/index.md

### Skipping workload cluster release

Skipping major versions when upgrading is unsupported. Upgrades must proceed sequentially through major versions.

You can skip minor and patch releases if needed. Our interfaces default to the next active workload cluster release. To skip a version:

- In the **GitOps** workflow, update the cluster configuration file with the desired release version.
- Using **kubectl-gs**, use the [`update cluster`]({{< relref "/reference/kubectl-gs/update-cluster" >}}) command with the `--release-version` flag.

## How upgrades work

Upgrades occur at runtime, maintaining workload functionality (with certain requirements). Key steps include:

- Nodes are drained, stopped, and recreated.
- Pods are rescheduled during node draining.
- Control plane node recreation could causes brief Kubernetes API unavailability though the API server is highly available minimizing the impact.

### Specific details for AWS

The control plane machines are rotated first, followed by worker nodes. The control plane nodes are replaced one by one, ensuring the cluster remains operational.

> The Cluster API for AWS controller manages clusters using EC2 instances. The default configuration of [instance warmup](https://github.com/search?q=repo%3Agiantswarm%2Fcluster-aws%20refreshPreferences&type=code) ensures AWS doesn't replace all nodes at once but in steps, allowing human intervention if needed. For a small node pool, one node is replaced every 10 minutes. For larger pools, small sets of nodes are replaced every 10 minutes.

Reviewer: ```suggestion
The Cluster API for AWS controller manages the cluster control plane using EC2 instances. The default configuration of [instance warmup](https://github.com/search?q=repo%3Agiantswarm%2Fcluster-aws%20refreshPreferences&type=code) ensures AWS doesn't replace all nodes at once but in steps, allowing human intervention if needed. For a small node pool, one node is replaced every 10 minutes. For larger pools, small sets of nodes are replaced every 10 minutes.
```

Once the worker nodes are rolled, each instance receives a terminate signal from AWS. Thanks to [`aws-node-termination-handler`](https://github.com/aws/aws-node-termination-handler) the machines are drained gracefully before being terminated. By default, the timeout configure to complete the draining operation is [`global.nodePools.PATTERN.awsNodeTerminationHandler.heartbeatTimeoutSeconds=1800`](https://github.com/giantswarm/cluster-aws/blob/main/helm/cluster-aws/README.md#node-pools). Customers can adjust this value to fit their needs in case they have specific requirements.

### Specific details for Azure

TBD

### Specific details for vSphere

TBD

### Specific details for vCloud Director

> TBD

Author: @vxav can you help me here to describe any peculiarity of those providers?

- [Cilium](https://cilium.io/) for networking and security
- [An observability bundle](https://github.com/giantswarm/observability-bundle)
- [A security bundle](https://github.com/giantswarm/security-bundle)

Each release is specific to the provider (AWS, Azure, or VMware) and identified by a version number. See the [workload cluster releases](https://github.com/giantswarm/releases/) repository for all available releases. Releases are **immutable** once deployed, ensuring consistency. Changes only happen through upgrades to a new release.

**Patch upgrades** increase the patch version for bug fixes and stability improvements.

> **Minor upgrades** introduce non-disruptive features that are typically less significant than major updates.

Reviewer 2: ```suggestion
**Minor upgrades** introduce non-disruptive features that are typically less significant than major upgrades.
```

**Major upgrades** include new Kubernetes minor releases, third-party components, and significant new features. Major releases align with Kubernetes upstream minor releases.

When a new major version is released, the oldest major release is deprecated. You can only create clusters with deprecated releases using `kubectl-gs`. Existing clusters remain unaffected.

## How to upgrade

**Patch and minor upgrades** are automatically rolled out by Giant Swarm, respecting agreed maintenance windows.

**Major upgrades** are not automated. You'll be notified to schedule the upgrade, possibly after updating workloads. Giant Swarm staff may assist or initiate these upgrades.

> You cannot skip major versions when upgrading - upgrades must proceed sequentially through major versions. You can skip minor and patch releases if needed.

Reviewer 2: ```suggestion
You cannot skip major versions when upgrading (e.g. v32.x.y to v34.x.y) – upgrades must proceed sequentially through major versions. You can usually skip minor and patch releases if needed. Please always check the [changelog]({{< relref "/changes" >}}).
```

To trigger an upgrade:

- **GitOps**: Update the cluster configuration file with the desired release version

> - **kubectl-gs**: Use the [`update cluster`]({{< relref "/reference/kubectl-gs/update-cluster" >}}) command with the `--release-version` flag

Reviewer 2: ```suggestion
- **kubectl-gs** command line tool: Use the [`update cluster`]({{< relref "/reference/kubectl-gs/update-cluster" >}}) command with the `--release-version` flag
```

## Avoid downtime during upgrades

To ensure your workloads remain operational during upgrades:

- **Ensure redundancy**: Use two or more replicas for all critical deployments
- **Distribute workloads**: Use [inter-pod anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) to spread replicas across nodes
- **Use probes**: Implement [liveness and readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) with appropriate `initialDelaySeconds` values
- **Handle termination signals**: Ensure pods handle `TERM` signals gracefully and configure `terminationGracePeriodSeconds` for longer shutdowns if needed
- **Set disruption budgets**: Use [PodDisruptionBudgets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) to maintain minimum pod counts (avoid `maxUnavailable=0` for single-replica deployments)
- **Set scheduling priorities**: Use Pod priority classes and set resource requests/limits to influence scheduling
- **Avoid ephemeral resources**: Use controllers (Deployments, StatefulSets) instead of standalone Pods

> - **Configure webhook timeouts**: Set low timeout values (e.g., 5 seconds) for validation and mutation webhooks
> - **Verify pod health**: Ensure all critical pods are running before upgrades with `kubectl get pods --all-namespaces | grep -v "Running\|ContainerCreating"`

Reviewer 2: ```suggestion
- **Configure webhook timeouts**: Set low timeout values (e.g., 5 seconds) for validation and mutation webhooks
- **Verify pod health**: Ensure all critical pods are running before upgrades with `kubectl get pods --all-namespaces | grep -v "Running\|Completed"`
```

`ContainerCreating` is similar to `Pending`, so only `Running` and `Completed` are _good_ signals.

## How upgrades work under the hood

Upgrades occur at runtime while maintaining workload functionality. Control plane nodes are recreated (causing temporary API slowness), then worker nodes are drained, stopped, and recreated while pods get rescheduled.

{{< tabs >}}
{{< tab id="aws-upgrade-details" for-impl="capa" title="AWS" >}}

Control plane machines are rotated first, followed by worker nodes. Control plane nodes are replaced one by one to keep the cluster operational.

The Cluster API for AWS controller manages upgrades using EC2 instances. The default [instance warmup](https://github.com/search?q=repo%3Agiantswarm%2Fcluster-aws%20refreshPreferences&type=code) configuration ensures AWS doesn't replace all nodes at once but in steps, allowing human intervention if needed. For small node pools, one node is replaced every 10 minutes. For larger pools, small sets of nodes are replaced every 10 minutes.

> Worker nodes receive a terminate signal from AWS. The [`aws-node-termination-handler`](https://github.com/aws/aws-node-termination-handler) drains machines gracefully before termination. The default timeout is [`global.nodePools.PATTERN.awsNodeTerminationHandler.heartbeatTimeoutSeconds=1800`](https://github.com/giantswarm/cluster-aws/blob/main/helm/cluster-aws/README.md#node-pools), which you can adjust as needed.

Reviewer 2: ```suggestion
Worker nodes receive a terminate signal from AWS. The preinstalled [`aws-node-termination-handler`](https://github.com/aws/aws-node-termination-handler) drains machines gracefully before termination. The default timeout is [`global.nodePools.PATTERN.awsNodeTerminationHandler.heartbeatTimeoutSeconds=1800`](https://github.com/giantswarm/cluster-aws/blob/main/helm/cluster-aws/README.md#node-pools), which you can adjust as needed.
```

{{< /tab >}}

> {{< tab id="azure-upgrade-details" for-impl="capz" title="Azure" >}}

Reviewer 2: Let's remove TBD tabs until we have any details there. Maybe a tab "Other providers" that says we don't have any details?!
